### PR TITLE
Add measurement/timer system and sourceFile to Statistics base class

### PR DESCRIPTION
**Pull request by haved**
The Statistics part of #382. Modifies all existing Statistics to provide the sourceFile path.

The mapping from id to name is now in `Statistics.cpp`, but many of the existing statistics used names that were different from the C++ enums, so I tried to make the classes keep their old names.

It might be better to have `GetName()` be an abstract method, and let subclasses provide names that way, instead of a single BijectiveMap. Let me know!

### CODE REVIEW

File: jlm/llvm/opt/InvariantValueRedirection.cpp [hunk 1]
```diff
@@ -19,8 +19,8 @@ class InvariantValueRedirection::Statistics final : public util::Statistics
 public:
   ~Statistics() override = default;
 
-  Statistics()
-      : util::Statistics(Statistics::Id::InvariantValueRedirection)
+  Statistics(util::filepath sourceFile)
```
Comment (phate) [inline jlm/llvm/opt/InvariantValueRedirection.cpp:22]: `explicit` would be nice.

File: jlm/llvm/opt/cne.cpp [hunk 2]
```diff
@@ -16,11 +16,10 @@ namespace jlm::llvm
 class cnestat final : public util::Statistics
 {
 public:
-  virtual ~cnestat()
-  {}
+  ~cnestat() override = default;
 
-  cnestat()
-      : Statistics(Statistics::Id::CommonNodeElimination),
+  cnestat(util::filepath sourceFile)
```
Comment (phate) [inline jlm/llvm/opt/cne.cpp:21]: `explicit`

File: jlm/llvm/opt/inlining.cpp [hunk 3]
```diff
@@ -16,11 +16,10 @@ namespace jlm::llvm
 class ilnstat final : public util::Statistics
 {
 public:
-  virtual ~ilnstat()
-  {}
+  ~ilnstat() override = default;
 
-  ilnstat()
-      : Statistics(Statistics::Id::FunctionInlining),
+  ilnstat(util::filepath sourceFile)
```
Comment (phate) [inline jlm/llvm/opt/inlining.cpp:21]: `explicit`

File: jlm/util/Statistics.hpp [hunk 4]
```diff
@@ -62,11 +74,123 @@ class Statistics
     return StatisticsId_;
   }
 
+  /**
+   * @return a string identifying the type of this Statistics instance
+   */
+  [[nodiscard]] std::string_view
```
Comment (phate) [inline jlm/util/Statistics.hpp:80]: `const std::string_view &` as return type?

File: jlm/util/Statistics.cpp [hunk 5]
```diff
@@ -1,15 +1,171 @@
...
+  return false;
+}
+
+Statistics::Measurement &
+Statistics::GetMeasurement(const std::string & name)
+{
+  for (auto & [mName, measurement] : Measurements_)
+    if (mName == name)
+      return measurement;
+  throw util::error(util::strfmt("Unknown measurement: ", name));
```
Comment (phate) [inline jlm/util/Statistics.cpp:115]: I would almost like to replace this with `JLM_ASSERT()`. Nobody catches it anyway, and it is basically a bug to ask for a measurement that is not present.

File: jlm/util/Statistics.cpp [hunk 6]
```diff
@@ -1,15 +1,171 @@
...
+  return false;
+}
+
+util::timer &
+Statistics::GetTimer(const std::string & name)
+{
+  for (auto & [mName, timer] : Timers_)
+    if (mName == name)
+      return timer;
+  throw util::error(util::strfmt("Unknown timer: ", name));
```
Comment (phate) [inline jlm/util/Statistics.cpp:145]: `JLM_ASSERT()`

File: jlm/util/Statistics.hpp [hunk 7]
```diff
@@ -62,11 +74,123 @@ class Statistics
...
+   * Retrieves the full list of timers
+   */
+  [[nodiscard]] util::iterator_range<TimerList::const_iterator>
+  GetTimers() const;
+
+protected:
+  /**
+   * Adds a measurement, identified by \p name, with the given value.
+   * Requires that the measurement doesn't already exist.
+   * Measurements are listed in insertion order.
```
Comment (phate) [inline jlm/util/Statistics.hpp:161]: This is not really relevant for the documentation of this method, right?

File: jlm/util/Statistics.hpp [hunk 8]
```diff
@@ -62,11 +74,123 @@ class Statistics
...
+   */
+  template<typename T>
+  void
+  AddMeasurement(std::string name, T value)
+  {
+    JLM_ASSERT(!HasMeasurement(name));
+    Measurements_.emplace_back(std::make_pair(std::move(name), std::move(value)));
+  }
+
+  // Mutable version of @see GetMeasurement
```
Comment (phate) [inline jlm/util/Statistics.hpp:172]: Does doxygen recognize `//`?

File: jlm/util/Statistics.hpp [hunk 9]
```diff
@@ -62,11 +74,123 @@ class Statistics
...
+
+  /**
+   * Creates a new timer with the given \p name.
+   * Requires that the timer does not already exist.
+   * @return a reference to the timer
+   */
+  util::timer &
+  AddTimer(std::string name);
+
+  // Mutable version of @see GetTimer
```
Comment (phate) [inline jlm/util/Statistics.hpp:184]: Does doxygen recognize `//`?

File: jlm/util/Statistics.hpp [hunk 10]
```diff
@@ -62,11 +74,123 @@ class Statistics
...
+  [[nodiscard]] util::timer &
+  GetTimer(const std::string & name);
 
 private:
   Statistics::Id StatisticsId_;
+  jlm::util::filepath SourceFile_;
```
Comment (phate) [inline jlm/util/Statistics.hpp:190]: `util::filepath SourceFile_`

File: tests/jlm/util/TestStatistics.cpp [hunk 11]
```diff
@@ -14,36 +14,58 @@
...
+  filepath path("file.ll");
+  MyTestStatistics statistics(Statistics::Id::Aggregation, path);
+
+  statistics.Start(10, 6.0);
+  statistics.Stop(-400, "poor");
+
+  assert(statistics.GetMeasurementValue<uint64_t>("count") == 10);
+  assert(statistics.GetMeasurementValue<double>("weight") == 6.0);
+  assert(statistics.GetMeasurementValue<int64_t>("bankAccount") == -400);
+  assert(statistics.GetMeasurementValue<std::string>("count") == "poor");
```
Comment (phate) [inline tests/jlm/util/TestStatistics.cpp:51]: I think the key should be `"state"` instead of `"count"`

**Review by phate — DISMISSED** [review]
Some minor issues, but nothing critical. Up to you whether you want to fix them.

I guess the next steps are (?):
1. Convert all statistics to use the new interface
2. Clean up the Serialize()/ToString() methods of the Statistics class

If you want, I can take care of the (mundane) first point such that you can continue working on Andersen.

**Comment by haved** [comment]
Thanks for the comments, I fixed almost all of them. I elected to go for `const util::filepath & sourceFile`, and remove a bunch of `std::move`.

I also modified the set of methods in `Statistics.hpp`, since it turns out C++ does overload resolution before checking visibility. Having public-but-const and protected non-const versions of the same method worked surprisingly poorly.

The only comment you had that I didn't act on was returning `string_view` vs. `const string_view &`, as I think the first is more flexible, and doesn't involve any copying.

Regarding converting classes, I was not planning on doing that work now. I want to get Andersen's statistics over to the `AddMeasurement` functionality, so that it becomes easier to add my new worklist implementation side-by-side.

**Review by phate — APPROVED** [review]
LGTM

**Pull request merged**